File: python/deixic_examples/account_brief.py
"""Prepare an account brief from CRM data available in your Deixic workspace."""

from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path

from console.v1 import console_pb2 as pb
from deixic import Deixic, DeixicError

from .task_result import save
from .account_brief_result import FORMAT_INSTRUCTION, parse_account_brief


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "action", choices=("check", "start", "resume", "replay", "approve", "deny")
    )
    parser.add_argument("state", type=Path, nargs="?")
    parser.add_argument("--channel")
    parser.add_argument("--account")
    parser.add_argument(
        "--structured",
        action="store_true",
        help="Request a validated JSON account brief",
    )
    parser.add_argument(
        "--progress",
        action="store_true",
        help="Write matching-turn event IDs to stderr",
    )
    parser.add_argument(
        "--request", help="Approval request ID selected by the authorized operator"
    )
    parser.add_argument(
        "--decision-key", help="Stable idempotency key for this explicit decision"
    )
    parser.add_argument(
        "--trigger", help="Stable ID for this business request; reused on recovery"
    )
    parser.add_argument(
        "--timeout", type=float, default=60, help="Observation budget in seconds"
    )
    args = parser.parse_args()
    if args.action != "check" and args.state is None:
        parser.error("This action requires a checkpoint path")
    if args.action == "start" and (not args.account or not args.trigger):
        parser.error("start requires --account and --trigger")
    if args.action in ("resume", "replay", "approve", "deny") and any(
        (args.account, args.trigger, args.channel, args.structured)
    ):
        parser.error(
            "This action uses the saved request; omit --account, --trigger, --channel and --structured"
        )
    if args.action in ("approve", "deny"):
        if not args.request or not args.decision_key:
            parser.error("approve and deny require --request and --decision-key")
    elif args.request or args.decision_key:
        parser.error("--request and --decision-key require approve or deny")
    if args.structured and args.action != "start":
        parser.error("--structured requires start; resume uses the saved format")
    try:
        required = ("DEIXIC_API_KEY", "DEIXIC_ORGANIZATION_ID", "DEIXIC_WORKSPACE_ID")
        missing = [name for name in required if not os.environ.get(name, "").strip()]
        if missing:
            raise ValueError("Missing configuration: " + ", ".join(missing))
        client = Deixic(
            api_key=os.environ["DEIXIC_API_KEY"],
            organization_id=os.environ["DEIXIC_ORGANIZATION_ID"],
            workspace_id=os.environ["DEIXIC_WORKSPACE_ID"],
            base_url=os.environ.get("DEIXIC_BASE_URL", "https://app.deixic.com"),
        )
        if args.action == "check":
            check = client.tasks.check_setup(channel_id=args.channel or "company")
            print(
                json.dumps(
                    dict(
                        status=check.status,
                        channel_id=check.channel_id,
                        write_access=check.write_access,
                        next_action=check.next_action,
                        requirements=[
                            dict(
                                service=item.service,
                                missing_requirements=list(item.missing_requirements),
                                reason_codes=[
                                    reason.reason_code
                                    for reason in item.missing_requirement_states
                                ],
                            )
                            for item in check.capabilities
                        ],
                        model_ready=check.selected_model.ready
                        if check.selected_model
                        else None,
                        error_kind=check.error.kind if check.error else None,
                        request_id=check.error.request_id if check.error else None,
                    )
                )
            )
            return 0 if check.status == "accessible" else 2

        if args.action == "start":
            # A new file is exclusive; subsequent writes atomically replace it.
            creating = True

            def persist(state):
                nonlocal creating
                save(args.state, state, create=creating)
                creating = False

            task = client.tasks.prepare(
                channel_id=args.channel or "company",
                body="Prepare an account brief for " + json.dumps(args.account) + ". "
                "Use CRM data available in this workspace. Include the account summary, "
                "open opportunities, risks and source references. State which data is missing. "
                "Do not update records or send messages."
                + (FORMAT_INSTRUCTION if args.structured else ""),
                idempotency_key=args.trigger,
                on_checkpoint=persist,
            )
            task.submit()
        else:
            task = client.tasks.resume(
                json.loads(args.state.read_text()),
                on_checkpoint=lambda state: save(args.state, state),
            )
            if args.action == "replay":
                task.replay()
            elif args.action in ("approve", "deny"):
                outcome = task.result()
                event = outcome.event
                if (
                    outcome.status != "waiting"
                    or event is None
                    or event.request_type != pb.OPERATING_THREAD_REQUEST_TYPE_APPROVAL
                    or event.request_id != args.request
                ):
                    print(
                        json.dumps(
                            dict(
                                status="needs_attention",
                                reason="approval_request_not_current",
                            )
                        )
                    )
                    return 2
                client.controls.respond(
                    channel_id=task.checkpoint()["channelId"],
                    turn_id=outcome.turn_id,
                    response=pb.OperatingThreadResponse(
                        request_id=event.request_id,
                        call_id=event.request_call_id,
                        request_type=event.request_type,
                        action=pb.OPERATING_THREAD_RESPONSE_ACTION_APPROVE
                        if args.action == "approve"
                        else pb.OPERATING_THREAD_RESPONSE_ACTION_DENY,
                    ),
                    idempotency_key=args.decision_key,
                )
                print(
                    json.dumps(
                        dict(
                            status="decision_submitted",
                            turn_id=outcome.turn_id,
                            next_action="resume the checkpoint to observe the owner outcome",
                        )
                    )
                )
                return 0
            else:

                def progress(event):
                    if args.progress:
                        print(
                            json.dumps(
                                dict(
                                    event_id=event.event_id,
                                    turn_id=event.turn_id,
                                    kind=event.kind,
                                )
                            ),
                            file=sys.stderr,
                            flush=True,
                        )

                outcome = task.wait(timeout=args.timeout, on_event=progress)
                structured = FORMAT_INSTRUCTION in task.checkpoint()["body"]
                brief = None
                if outcome.status == "completed" and structured:
                    try:
                        brief = outcome.parse(parse_account_brief)
                        expected, _ = json.JSONDecoder().raw_decode(
                            task.checkpoint()["body"][
                                len("Prepare an account brief for ") :
                            ]
                        )
                        if brief["accountName"] != expected:
                            raise ValueError("Account mismatch")
                    except (ValueError, TypeError):
                        print(
                            json.dumps(
                                dict(
                                    status="invalid_result",
                                    remote_status="completed",
                                    turn_id=outcome.turn_id,
                                    next_action="inspect the answer; do not resubmit this request",
                                )
                            )
                        )
                        return 2
                actions = [
                    dict(
                        id=item.id,
                        owner_service=item.owner_service,
                        object_id=item.object_id,
                        kind=item.kind,
                        lifecycle_state=item.lifecycle_state,
                        evidence_refs=[
                            dict(resource_type=ref.resource_type, id=ref.id)
                            for ref in item.evidence_refs
                        ],
                    )
                    for item in outcome.receipts
                ]
                attention = any(
                    item.lifecycle_state
                    not in (
                        pb.RECEIPT_LIFECYCLE_STATE_SUCCEEDED,
                        pb.RECEIPT_LIFECYCLE_STATE_VERIFIED,
                    )
                    for item in outcome.receipts
                )
                print(
                    json.dumps(
                        dict(
                            status=outcome.status,
                            turn_id=outcome.turn_id,
                            reason=outcome.reason,
                            body=outcome.body if not structured else None,
                            brief=brief,
                            receipt_ids=[item.id for item in outcome.receipts],
                            actions=actions,
                            action_status="not_reported"
                            if not actions
                            else "requires_attention"
                            if attention
                            else "owner_reported_success",
                            request_id=outcome.event.request_id
                            if outcome.event
                            else None,
                            request_type=outcome.event.request_type
                            if outcome.event
                            else None,
                            error_code=outcome.turn.error_code
                            if outcome.turn
                            else None,
                        )
                    )
                )
                return 0 if outcome.status == "completed" and not attention else 2
        print(
            json.dumps(
                dict(
                    status="accepted",
                    turn_id=task.checkpoint()["turnId"],
                    next_action="resume this checkpoint to retrieve the account brief",
                )
            )
        )
        return 0
    except DeixicError as error:
        print(
            json.dumps(
                dict(
                    status="error",
                    kind=error.kind,
                    code=error.code,
                    request_id=error.request_id,
                )
            )
        )
        return 1
    except (ValueError, OSError) as error:
        # Only our missing-configuration diagnostic is safe to print verbatim.
        message = (
            str(error)
            if str(error).startswith("Missing configuration:")
            else "Check configuration and checkpoint storage"
        )
        print(
            json.dumps(dict(status="error", kind="configuration", next_action=message))
        )
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/account-brief.mjs
#!/usr/bin/env node
import { open, readFile, rename, link, mkdtemp, rm } from "node:fs/promises";
import { basename, dirname, join } from "node:path";
import { parseArgs } from "node:util";
import { createDeixicClient, DeixicError, parseTaskResult, OperatingThreadRequestType, OperatingThreadResponseAction, ReceiptLifecycleState } from "@evalops/deixic-sdk";

import { FORMAT_INSTRUCTION, parseAccountBrief } from "./account-brief-result.mjs";

async function save(path, state, create = false) {
  const directory = await mkdtemp(join(dirname(path), basename(path) + "."));
  const target = join(directory, "checkpoint.json");
  try {
    const file = await open(target, "wx", 0o600);
    try { await file.writeFile(JSON.stringify(state)); await file.sync(); }
    finally { await file.close(); }
    if (create) await link(target, path);
    else await rename(target, path);
    const parent = await open(dirname(path), "r");
    try { await parent.sync(); } finally { await parent.close(); }
  } finally { if (directory) await rm(directory, { recursive: true, force: true }); }
}

async function main() {
  if (process.argv.includes("--help")) {
    console.log("deixic-account-brief check [--channel ID]\n"
      + "deixic-account-brief start CHECKPOINT --account NAME --trigger STABLE_ID [--channel ID]\n"
      + "deixic-account-brief resume CHECKPOINT [--timeout MILLISECONDS] [--progress]\n"
      + "deixic-account-brief approve|deny CHECKPOINT --request ID --decision-key STABLE_ID\n"
      + "Add --structured to start to request and validate a JSON account brief.\n"
      + "deixic-account-brief replay CHECKPOINT\n"
      + "Uses CRM data available in your Deixic workspace. Set DEIXIC_API_KEY, DEIXIC_ORGANIZATION_ID and DEIXIC_WORKSPACE_ID.");
    return 0;
  }
  const { positionals, values } = parseArgs({ allowPositionals: true, options: {
    channel: { type: "string" }, account: { type: "string" }, trigger: { type: "string" },
    timeout: { type: "string", default: "60000" },
    structured: { type: "boolean", default: false }, progress: { type: "boolean", default: false },
    request: { type: "string" }, "decision-key": { type: "string" },
  } });
  const [action, path] = positionals;
  if (!["check", "start", "resume", "replay", "approve", "deny"].includes(action) || positionals.length > 2
    || action !== "check" && !path) throw new Error("configuration");
  if (action === "start" && (!values.account || !values.trigger)) throw new Error("configuration");
  if (["resume", "replay", "approve", "deny"].includes(action) && (values.account || values.trigger || values.channel || values.structured)) throw new Error("configuration");
  if (["approve", "deny"].includes(action) ? (!values.request || !values["decision-key"])
    : (values.request || values["decision-key"])) throw new Error("configuration");
  if (values.structured && action !== "start") throw new Error("configuration");
  const required = ["DEIXIC_API_KEY", "DEIXIC_ORGANIZATION_ID", "DEIXIC_WORKSPACE_ID"];
  const missing = required.filter(name => !process.env[name]?.trim());
  if (missing.length) {
    console.log(JSON.stringify({ status: "error", kind: "configuration", nextAction: "Missing configuration: " + missing.join(", ") }));
    return 1;
  }
  const client = createDeixicClient({ apiKey: process.env.DEIXIC_API_KEY,
    organizationId: process.env.DEIXIC_ORGANIZATION_ID, workspaceId: process.env.DEIXIC_WORKSPACE_ID,
    baseUrl: process.env.DEIXIC_BASE_URL,
  });
  if (action === "check") {
    const check = await client.tasks.checkSetup({ channelId: values.channel ?? "company" });
    console.log(JSON.stringify({ status: check.status, channelId: check.channelId,
      writeAccess: check.writeAccess, nextAction: check.nextAction,
      requirements: check.capabilities.map(item => ({ service: item.service,
        missingRequirements: item.missingRequirements,
        reasonCodes: item.missingRequirementStates.map(reason => reason.reasonCode) })),
      modelReady: check.selectedModel?.ready ?? null, errorKind: check.error?.kind,
      requestId: check.error?.requestId,
    }));
    return check.status === "accessible" ? 0 : 2;
  }
  let task;
  if (action === "start") {
    let creating = true;
    task = await client.tasks.prepare({ channelId: values.channel ?? "company",
      body: "Prepare an account brief for " + JSON.stringify(values.account) + ". "
        + "Use CRM data available in this workspace. Include the account summary, "
        + "open opportunities, risks and source references. State which data is missing. "
        + "Do not update records or send messages." + (values.structured ? FORMAT_INSTRUCTION : ""),
      idempotencyKey: values.trigger, onCheckpoint: async state => {
        await save(path, state, creating); creating = false;
      },
    });
    await task.submit();
  } else {
    task = client.tasks.resume(JSON.parse(await readFile(path, "utf8")), { onCheckpoint: state => save(path, state) });
    if (action === "replay") await task.replay();
    else if (["approve", "deny"].includes(action)) {
      const result = await task.result();
      const event = result.event;
      if (result.status !== "waiting" || !event || event.requestType !== OperatingThreadRequestType.APPROVAL
        || event.requestId !== values.request) {
        console.log(JSON.stringify({ status: "needs_attention", reason: "approval_request_not_current" }));
        return 2;
      }
      await client.controls.respond({ channelId: task.checkpoint().channelId, turnId: result.turnId,
        response: { requestId: event.requestId, callId: event.requestCallId, requestType: event.requestType,
          action: action === "approve" ? OperatingThreadResponseAction.APPROVE : OperatingThreadResponseAction.DENY },
        idempotencyKey: values["decision-key"],
      });
      console.log(JSON.stringify({ status: "decision_submitted", turnId: result.turnId,
        nextAction: "resume the checkpoint to observe the owner outcome" }));
      return 0;
    } else {
      const result = await task.wait({ timeoutMs: Number(values.timeout), onEvent: event => {
        if (values.progress) process.stderr.write(JSON.stringify({ eventId: event.eventId,
          turnId: event.turnId, kind: event.kind }) + "\n");
      } });
      const structured = task.checkpoint().body.includes(FORMAT_INSTRUCTION);
      let brief;
      if (result.status === "completed" && structured) {
        try {
          brief = parseTaskResult(result, parseAccountBrief);
          const name = task.checkpoint().body.slice("Prepare an account brief for ".length).match(/^"(?:\\.|[^"\\])*"/);
          if (!name || brief.accountName !== JSON.parse(name[0])) throw new Error("Account mismatch");
        }
        catch {
          console.log(JSON.stringify({ status: "invalid_result", remoteStatus: "completed", turnId: result.turnId,
            nextAction: "inspect the answer; do not resubmit this request" }));
          return 2;
        }
      }
      const receipts = result.status === "completed" ? result.receipts : [];
      const actions = receipts.map(item => ({ id: item.id, ownerService: item.ownerService,
        objectId: item.objectId, kind: item.kind, lifecycleState: item.lifecycleState,
        evidenceRefs: item.evidenceRefs.map(ref => ({ resourceType: ref.resourceType, id: ref.id })) }));
      const attention = receipts.some(item => ![ReceiptLifecycleState.SUCCEEDED, ReceiptLifecycleState.VERIFIED].includes(item.lifecycleState));
      console.log(JSON.stringify({ status: result.status, turnId: result.turnId,
        reason: result.reason, body: result.status === "completed" && !structured ? result.body : undefined,
        brief, receiptIds: receipts.map(item => item.id), actions,
        actionStatus: !actions.length ? "not_reported" : attention ? "requires_attention" : "owner_reported_success",
        requestId: result.event?.requestId, requestType: result.event?.requestType, errorCode: result.turn?.errorCode,
      }));
      return result.status === "completed" && !attention ? 0 : 2;
    }
  }
  console.log(JSON.stringify({ status: "accepted", turnId: task.checkpoint().turnId,
    nextAction: "resume this checkpoint to retrieve the account brief" }));
  return 0;
}

try { process.exitCode = await main(); }
catch (error) {
  console.log(JSON.stringify(error instanceof DeixicError
    ? { status: "error", kind: error.kind, code: error.code, requestId: error.requestId }
    : { status: "error", kind: "configuration", nextAction: "Check configuration and checkpoint storage; use --help" }));
  process.exitCode = 1;
}
